File: C++/6198_옥상정원꾸미기.cpp
#include <iostream>
#include <stack>
using namespace std;

int main() {
	int N;
	cin >> N;
	
	long long int height[N];
	for(int i = 0; i < N; i++){
		cin >> height[i];
	} // 높이를 담을 배열(높이 범위 주의! int 안됨)
	
	int num[N] = {0,}; // 각 건물이 볼 수 있는 건물들의 수
	
	stack<int> s; // 건물들의 index를 담을 stack
	s.push(N-1);
	for(int i = N-2; i >= 0; i--){ // 맨 끝의 건물은 볼 수 있는 건물이 없으므로 N-1번째 건물부터 확인하면서 앞으로 이동
		int cnt = 0;
		while(!s.empty()){ // stack이 비어있지 않아야 함
			if(height[s.top()] >= height[i]){ // 현재 stack에 있는 건물들의 높이 확인 더 높은 건물이라면 볼 수 없기 때문에 break
				break;
			}
			cnt++; // 현재 stack 맨 위의 건물은 볼 수 있는 건물이므로 볼 수 있는 건물 개수 1 추가
			
			int n = s.top();
			num[i] += num[n]; // 현재 stack 맨 위의 건물이 볼 수 있는 건물은 현재 건물도 볼 수 있는 건물이므로 모두 더해주기(이전에 pop 되어서 stack 안에는 존재하지 않을 것 -> stack 내에 없는 건물 중 현재 건물에서 볼 수 있는 건물의 수 추가) 
			s.pop();
		}
		num[i] += cnt; // 볼 수 있는 건물의 수 더해주기(stack 내에 있었던 건물 중 현재 있는 건물에서 볼 수 있는 건물의 수)
		
		s.push(i); // 확인한 건물의 index stack에 삽입
	}
	
	long long int result = 0; // result 범위 주의(int 안됨)
	for(int i = 0; i < N; i++){ // num 배열의 합 구하기
		result += num[i];
	}
	
	cout << result;
	
	return 0;
}

File: Python/6198_옥상정원꾸미기.py
if __name__ == '__main__':
	N = int(input())
	height = [int(input()) for _ in range(N)] # 건물들의 높이
	
	num = [0]*N # 현재 건물 위치에서 볼 수 있는 건물들의 수를 담을 배열
	stack = [N-1]
	for i in range(N-2,-1,-1):
		cnt = 0
		while stack and height[stack[-1]] < height[i]: # stack이 비어있지 않아야 하며, stack의 맨 위(top)의 건물 높이가 현재 건물 높이보다 낮아야 함
			cnt += 1 # 현재 건물보다 낮은 건물 개수 세기(stack 내에 존재하는 건물들 중 현재 건물보다 낮은 건물들의 개수 구하기)
			num[i] += num[stack.pop()] # 현재 건물보다 낮은 건물들에서 볼 수 있는 건물들의 개수 더하기(stack 내에 존재하지 않는 건물들 중 현재 건물보다 낮은 건물들의 개수 구하기)
			
		num[i] += cnt # stack 내에 존재하는 낮은 건물의 수
		
		stack.append(i) # 모두 계산 후 현재 건물의 index를 stack에 추가
        
	print(sum(num)) # 각 위치에서 볼 수 있는 건물들의 수 모두 더하기
